internal/wasi/fs.go

```diff
@@ -5,100 +5,137 @@ import (
 	"io"
 	"io/fs"
 	"os"
+	"path"
 	"runtime"
+	"sort"
 	"strings"
-
-	"github.com/tetratelabs/wazero/wasi"
+	"time"
 )
 
+// DirFS is a file system that implements fs.FS and wasi.OpenFileFS.
+// It is similar to os.DirFS, but supports file creation by OpenFile.
+// See os.DirFS.
 type DirFS string
 
-func posixOpenFlags(oFlags uint32, fsRights uint64) (pFlags int) {
-	if fsRights&wasi.R_FD_WRITE != 0 {
-		pFlags |= os.O_RDWR
-	}
-	if oFlags&wasi.O_CREATE != 0 {
-		pFlags |= os.O_CREATE
-	}
-	if oFlags&wasi.O_EXCL != 0 {
-		pFlags |= os.O_EXCL
+// OpenFile implements wasi.OpenFileFS.OpenFile.
+func (dir DirFS) OpenFile(name string, flag int, perm os.FileMode) (fs.File, error) {
+	if !fs.ValidPath(name) || runtime.GOOS == "windows" && strings.ContainsAny(name, `\:`) {
```

Comment:
I would avoid doing `runtime.GOOS == "windows"` everytime `OpenFile` is called.. one idea is to have `const isWindows = runtime.GOOS == "windows"` and use it here (and eventually it will be optimized by compiler as it's const.

Reply:
During changing it, I've realized that `runtime.GOOS` is compile-time constant string in Golang, not a value obtained at runtime. And `runtime.GOOS == "windows"` is completely optimized out and disappears on non-Windows binaries, so it never runs at runtime. So, I think it would be better to leave it as is this time, rather than introducing build flags only for this. wdyt?

e.g.
on Windows, it calls `strings.IndexAny` (= `ContainsAny`) after `ValidPath`. 
```asm
0000000000591ec0 <github.com/tetratelabs/wazero/internal/wasi.DirFS.OpenWASI>:
  --snip--
  591f02:   e8 39 ff ee ff          callq  481e40 <io/fs.ValidPath>
  591f07:   84 c0                   test   %al,%al
  591f09:   74 24                   je     591f2f <github.com/tetratelabs/wazero/internal/wasi.DirFS.OpenWASI+0x6f>
  591f0b:   48 8b 44 24 70          mov    0x70(%rsp),%rax
  591f10:   48 8b 5c 24 40          mov    0x40(%rsp),%rbx
  591f15:   48 8d 0d 5e b8 05 00    lea    0x5b85e(%rip),%rcx        # 5ed77a <go.string.*+0x222>
  591f1c:   bf 02 00 00 00          mov    $0x2,%edi
  591f21:   e8 da b8 f1 ff          callq  4ad800 <strings.IndexAny>
--snip--
```

on Linux, that path disappears.
```asm
0000000000583440 <github.com/tetratelabs/wazero/internal/wasi.DirFS.OpenWASI>:
--snip--
  583482:   e8 99 28 ef ff          callq  475d20 <io/fs.ValidPath>
  583487:   84 c0                   test   %al,%al
  583489:   0f 84 e2 00 00 00       je     583571 <github.com/tetratelabs/wazero/internal/wasi.DirFS.OpenWASI+0x131>
  58348f:   48 8b 94 24 88 00 00    mov    0x88(%rsp),%rdx
  583496:   00
  583497:   0f ba e2 09             bt     $0x9,%edx
  58349b:   73 17                   jae    5834b4 <github.com/tetratelabs/wazero/internal/wasi.DirFS.OpenWASI+0x74>
  58349d:   48 f7 c2 08 00 00 00    test   $0x8,%rdx
  5834a4:   be 01 00 00 00          mov    $0x1,%esi
  5834a9:   ba 02 00 00 00          mov    $0x2,%edx
  5834ae:   48 0f 45 f2             cmovne %rdx,%rsi
```

btw, it looks the this line originally comes from Golang's standard library. https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/os/file.go;l=644

Reply:
oh that's good to know! thank you for confirming that